lib/logstash/inputs/beats.rb

```diff
@@ -207,9 +210,229 @@ def invoke(connection, &block)
   # this thread take the element from one queue into another one.
   def start_buffer_broker
     @threadpool.post do
-      while !stop?
-        @output_queue << @buffered_queue.pop_no_timeout
+      LogStash::Util.set_thread_name("[beats-input]-buffered-queue-broker")
+
+      begin
+        while !stop?
+          @output_queue << @buffered_queue.take
+        end
+      rescue => e
+        # If we are shutting down without waiting the queue to unblock
+        # we will get an `InterruptionException` in that context we will not log it.
+        @logger.error("Beats input: bufferered queue broker exception", :exception => e) unless stop?
       end
     end
   end
+
+  # Handle the data coming from a connection
+  # Decide which Process should be used to decode the data coming
+  # from the beat library.
+  #
+  # - Should we use a codec on specific field?
+  # - Should we just take the raw content of the parsed json frame
+  class ConnectionHandler
+    def initialize(connection, input, queue)
+      @connection = connection
+
+      @input = input
+      @queue = queue
+      @logger = input.logger
+
+      # We need to clone the codec per connection, so we can flush a specific 
+      # codec when a connection is closed.
+      @codec = input.codec.dup
+      
+      @json_transformer = JsonTransform.new(@input)
+      @codec_transformer = CodecTransform.new(@input)
```

Comment:
> from what I understand, these are 2 transformers for the scenarios where the codec was and wasn't applied, right?

Correct, since the logic is kinda different in both case.

> Not a big fan of this name :(
The variable, the classes or both? Any suggestion ? :(